src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ReadOnlyTagHelperAttributes.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    public class ReadOnlyTagHelperAttributes<TAttributeType> : IReadOnlyList<TAttributeType>
+        where TAttributeType : IReadOnlyTagHelperAttribute
+    {
+        protected readonly List<TAttributeType> _attributes;
+
+        protected ReadOnlyTagHelperAttributes()
+        {
+            _attributes = new List<TAttributeType>();
+        }
+
+        public ReadOnlyTagHelperAttributes([NotNull] IEnumerable<TAttributeType> attributes)
+        {
+            _attributes = new List<TAttributeType>(attributes);
+        }
+
+        public TAttributeType this[int index]
+        {
+            get
+            {
+                return _attributes[index];
+            }
+        }
+
+        public TAttributeType this[[NotNull] string key]
+        {
+            get
+            {
+                return _attributes.FirstOrDefault(attribute => KeyEquals(key, attribute));
+            }
+        }
+
+        public int Count
+        {
+            get
+            {
+                return _attributes.Count;
+            }
+        }
+
+        public virtual bool IsReadOnly
+        {
+            get
+            {
+                return true;
+            }
+        }
+
+        public IEnumerable<string> Keys
+        {
+            get
+            {
+                return _attributes.Select(attribute => attribute.Name);
+            }
+        }
+
+        public IEnumerable<object> Values
+        {
+            get
+            {
+                return _attributes.Select(attribute => attribute.Value);
+            }
+        }
```

Comment:
If you're looking at it from the 5% case (multiple attribute same name) I completely agree. However there's more value here for the average case when keys are different.

As for the `Values`, I agree, it may not be needed. I can be swayed one way or the other. It only exists because `Keys` exists.